x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/regression/MeanSquaredError.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ml.dataframe.evaluation.regression;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregationBuilders;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.metrics.NumericMetricsAggregation;
+import org.elasticsearch.xpack.core.ml.dataframe.evaluation.EvaluationMetricResult;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Calculates the mean squared error between two known numerical fields.
+ *
+ * equation: mse = 1/n * Σ(y - y´)^2
+ */
+public class MeanSquaredError implements RegressionMetric {
+
+    public static final ParseField NAME = new ParseField("mean_squared_error");
+
+    private static final String PAINLESS_TEMPLATE = "def diff = doc[''{0}''].value - doc[''{1}''].value;return diff * diff;";
```

Comment:
@tveasey if we add them as "metrics" under the "regression" evaluation, R-Squared and MAE would be part of the same query. With how things are phased queries + aggs, they are applied at the "same time". We may "hit the same doc twice", but it would already be loaded on the shard. The resource utilization difference would be minuscule for the added complexity.

Reply:
Let me clarify, a `Regression` evaluation can have numerous metrics (characterized by unique aggs) but all are done in a single query.

Reply:
Ok, cool. I'd missed this detail: I was thinking each metric was responsible for actually performing it's own search. In that case, I think the main comment is "is it worth getting R^2 at the same time?". This is an interesting sort of metric because it can essentially be got from mean squared error together with variance for the actuals. This is a useful metric in its own right, but also incorporating this from the start it will be interesting to see how it fits in without code duplication.

Reply:
@tveasey sure, I can add it :).

Reply:
@tveasey ok, looking at what is possible with aggs, R^2 will be a two phase thing. We don't have the infrastructure in place for the evaluation API to do two phase metrics. This is something we can add in the future, but will definitely blow up the line count in this PR. 

I can add MAE instead if you would like.

Reply:
Ben an I discussed this a bit further offline. Computing r^2 is in fact possible without 2-phase, but we feel like it is probably worth delaying this to a separate PR since this is already quite large. We also discussed a separate thought which is should we have a layer which is responsible for gathering simple statistics which are feed into evaluation metrics, MSE and R^2 being examples of metrics which can reuse the same simple statistics. We'll discuss this with @dimitris-athanasiou when he's back.

Reply:
@tveasey  and I talked offline. He taught me that R^2 can be calculated utilizing the variance, so access to the mean directly is not strictly necessary.  :). I think adding new metrics should be booted to another PR to keep this size down.